src/usermap.rs

```diff
@@ -0,0 +1,214 @@
+use std::sync::{Arc, Mutex};
+
+use crate::event_emitter::EventEmitter;
+use crate::memcmp::{get_non_idle_user_filter, get_user_filter};
+use crate::utils::{decode, get_ws_url};
+use crate::websocket_program_account_subscriber::{
+    ProgramAccountUpdate, WebsocketProgramAccountOptions, WebsocketProgramAccountSubscriber,
+};
+use crate::SdkResult;
+use anchor_lang::AccountDeserialize;
+use dashmap::DashMap;
+use drift::state::user::User;
+use serde_json::json;
+use solana_account_decoder::UiAccountEncoding;
+use solana_client::nonblocking::rpc_client::RpcClient;
+use solana_client::rpc_config::{RpcAccountInfoConfig, RpcProgramAccountsConfig};
+use solana_client::rpc_request::RpcRequest;
+use solana_client::rpc_response::{OptionalContext, RpcKeyedAccount};
+use solana_sdk::commitment_config::CommitmentConfig;
+use solana_sdk::pubkey::Pubkey;
+
+pub struct Usermap {
+    subscribed: bool,
+    subscription: WebsocketProgramAccountSubscriber,
+    usermap: Arc<DashMap<String, User>>,
+    sync_lock: Option<Mutex<()>>,
+    latest_slot: Arc<AtomicU64>,
+    commitment: CommitmentConfig,
+    rpc: RpcClient,
+}
+
+impl Usermap {
+    pub fn new(commitment: CommitmentConfig, endpoint: String, sync: bool) -> Self {
+        let filters = vec![get_user_filter(), get_non_idle_user_filter()];
+        let options = WebsocketProgramAccountOptions {
+            filters,
+            commitment,
+            encoding: UiAccountEncoding::Base64,
+        };
+        let event_emitter = EventEmitter::new();
+
+        let url = get_ws_url(&endpoint.clone()).unwrap();
+
+        let subscription =
+            WebsocketProgramAccountSubscriber::new("usermap", url, options, event_emitter);
+
+        let usermap = Arc::new(DashMap::new());
+
+        let rpc = RpcClient::new_with_commitment(endpoint.clone(), commitment);
+
+        let sync_lock = if sync { Some(Mutex::new(())) } else { None };
+
+        Self {
+            subscribed: false,
+            subscription,
+            usermap,
+            sync_lock,
+            latest_slot: Arc::new(AtomicU64::new(0)),
+            commitment,
+            rpc,
+        }
+    }
+
+    pub async fn subscribe(&mut self) -> SdkResult<()> {
+        if let Some(_) = self.sync_lock {
+            self.sync().await?;
+        }
+
+        if !self.subscribed {
+            self.subscription.subscribe::<User>().await?;
+            self.subscribed = true;
+        }
+
+        let usermap = self.usermap.clone();
+        let latest_slot = self.latest_slot.clone();
+
+        self.subscription
+            .event_emitter
+            .subscribe("usermap", move |event| {
+                if let Some(update) = event.as_any().downcast_ref::<ProgramAccountUpdate<User>>() {
+                    let user_data_and_slot = update.data_and_slot.clone();
+                    let user_pubkey = update.pubkey.to_string();
+                    if update.data_and_slot.slot > latest_slot.load(Ordering::Relaxed) {
```

Comment:
maybe `>=` ? so updates from different user in the same slot should both go through

Reply:
the data updates will still go through, it just won't go to the effort of updating the latest slot if it's the same slot